### PR TITLE
Added an ability to authenticate requests to servers with jwtproxy

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?

In this PR secure servers are introduced. If server supposed to be secure it should contain `secure: true` attribute, for example ws-agent server configuration:
```json
    "wsagent/http": {
      "port": "4401/tcp",
      "protocol": "http",
      "path" : "/api",
      "attributes": {
        "secure": "true"
      }
    }
```
It is needed to make servers authentication independent.

Then infrastructure should cover server with authentication and server will receive only authenticated requests. Requests to secure servers must contain machine token in cookies, `Authorization` header or `token` query parameter.

This PR contains secure servers supporting of Kubernetes and OpenShift infrastructures. 

For some time, it is not enabled by default, so secure servers still should authenticate requests themselves. It may be enabled by setting `che.agents.auth.secure_server_exposer=jwtproxy`.

#### Kubernetes and OpenShift infrastructures implentation notes

It is done with JWTProxy, that is add in each Workspace as separated Pod. Original servers endpoints won't be accessible externally, instead JWTProxy port will be exposed and authenticated requests will be proxied to original server. 
The same components works for Kubernetes and OpenShift infrasturcture since they reuse ExternalExposeServerStrategy (for K8s - ingress based, for OS - route based) for exposing JWTProxy service's ports.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/10081

#### Docs PR
I think Docs should be added when this feature won't be beta

### CODE REVIEW

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java [hunk 1]
```diff
@@ -129,35 +134,68 @@ public KubernetesServerExposer(
...
     servers.forEach(
         (key, value) -> {
           if ("true".equals(value.getAttributes().get(INTERNAL_SERVER_ATTRIBUTE))) {
+            // Server is internal. It doesn't make sense to make an it secure since
+            // it is available only within workspace servers
             internalServers.put(key, value);
           } else {
-            externalServers.put(key, value);
+            // Server is external. Check if it should be secure or not
+            if ("true".equals(value.getAttributes().get(ServerConfig.SECURE_SERVER_ATTRIBUTE))) {
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java:150]: Boolean.parseBoolean(value.getAttributes().get(ServerConfig.SECURE_SERVER_ATTRIBUTE)) ?

**Review by skabashnyuk — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java (hunk 1 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java:150]: Thanks for pointing. And it was implemented in this way for internal servers. I think it is done in this way because comparing with `true` obviously shows that only this value is suitable here. Also, it is documented [here](https://github.com/eclipse/che/blob/2c2bd6637c8169018e2b33397f9d738bbf82a0b1/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/config/ServerConfig.java#L26-L27) that only `true` will be considered as right value for internal/secure servers. While `parseBoolean` checks value with `equalsIgnoreCase`. I'd leave it as it is, at least for now.

**Review by ghost — APPROVED** [review]

**Comment by sleshchenko** [comment]
@garagatyi @mshaposhnik @skabashnyuk Now PR is ready to review, please take a look.

**Comment by sleshchenko** [comment]
ci-test

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyConfigBuilder.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Helps to build JWTProxy config with several verifier proxies.
+ *
+ * @author Sergii Leshchenko
+ */
+public class JwtProxyConfigBuilder {
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyConfigBuilder.java:24]: This doesn't seem related to k8s. Maybe it worth to move it to another module?

**Comment by garagatyi** [comment]
Is this PR working or just a part of working new security system?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalInfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalMachineConfig;
+import org.eclipse.che.multiuser.machine.authentication.server.signature.SignatureKeyManager;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.ServerServiceBuilder;
+
+/** @author Sergii Leshchenko */
+public class JwtProxyProvisioner {
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java]: Please, add javadocs

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxySecureServerExposer.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.eclipse.che.api.core.model.workspace.config.ServerConfig;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.multiuser.machine.authentication.server.signature.SignatureKeyManager;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.external.ExternalServerExposerStrategy;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.secure.SecureServerExposer;
+
+/**
+ * Expose secure servers with JWTProxy.
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxySecureServerExposer.java]: Can you add more details into the javadocs? It would be helpful to understand what the class does and what for.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+   */
+  public ServicePort expose(
+      KubernetesEnvironment k8sEnv,
+      String backendServiceName,
+      int backendServicePort,
+      String protocol)
+      throws InfrastructureException {
+    ensureJwtProxyInjected(k8sEnv);
+
+    int listenPort = availablePort++;
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java:119]: Didn't get some use-cases. Would you be willing to explain them to me?
1. What if 2 workspace masters are running at the same time?
1. What if workspaces get deployed to the same single k8s namespace?
1. Is there any guarantee that this class won't be created multiple times since it is not a singleton and is not created automatically by DI container?

**Review by garagatyi — COMMENTED** [review]
Overall looks good! There are inlined comments about the things I would love to understand more/address in the PR to provide a meaningful review.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+  private final boolean agentsAuthEnabled;
+  private final String serverExposer;
+
+  private final ExternalServerExposerStrategy<T> exposerStrategy;
+  private final JwtProxySecureServerExposerFactory<T> jwtProxySecureServerExposerFactory;
+
+  @Inject
+  public SecureServerExposerFactory(
+      @Named("che.agents.auth_enabled") boolean agentsAuthEnabled,
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: Thank, Good point! It makes more sense to use `servers` word there instead of `agents`. I will change it.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyConfigBuilder.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  public String build() {
+    StringBuilder configBuilder = new StringBuilder();
+
+    configBuilder.append("jwtproxy:\n" + "  verifier_proxies:\n");
+    for (VerifierProxy verifierProxy : verifierProxies) {
+      configBuilder.append(
+          String.format(
+              "  - listen_addr: :%s\n" // :4471
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyConfigBuilder.java:43]: I've looked at this approach but it did not make code easier since I need to create 4 or even more additionally classes. I'll try again, maybe it would be better now =)

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyConfigBuilder.java (hunk 2 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyConfigBuilder.java:24]: Good point, but I have no idea which module will be good to place for this class. I would leave it here while it is used by Kubernetes infrastructure. And move it to a new separated module or existing one when there will be a case where JwtProxyConfigBuilder will be used in another module, like Docker infrastructure. Is it OK for you?

Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyConfigBuilder.java:24]: Sure

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java (hunk 5 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java:119]: > 1. What if 2 workspace masters are running at the same time?

There is no issue with that because of a restriction that only 1 Che Workspace Master is able to start a particular workspace at the same time.

> 2. What if workspaces get deployed to the same single k8s namespace?

There are no any issues with that while it is a port of a jwtproxy pod for a particular workspace. Jwtproxy pod will have a unique name within k8s namespace that is set by `UniqueNamesProvisioner` class.

> 3. Is there any guarantee that this class won't be created multiple times since it is not a singleton and is not created automatically by DI container?

Actually, this class MUST BE created once for one `STARTING` workspace, I don't see any possible ways in which we can guarantee that class is used correctly. What we can do - describe in java docs how it should be used.

One more thing that we could implement to prevent conflicts while incorrect usage of this class - use unique objects name - like machine name, pod name, etc. Then if this class will be created two or more times then the only thing that happens - there are more jwtproxy pods that are needed but not ports, objects conflicts should be there and everything would work as expected.
But I think it may be a bit overkill while this class is supposed to use in one place.

One more note, about unique names - do you think should we use a unique machine name instead of hardcoded? Now we always use `jwtproxy` machine name. It would be better to use generated name with prefix like `jwtproxy9sjXnw6y` to avoid conflicts when user have jwtproxy name?

Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java:119]: ok, I got it. Yeah, the current solution is good enough, overcomplicating it is not needed. Thanks for the clarification!

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: @garagatyi Do you think it may be `che.server.secure_exposer` property which is Che System scope or it would be better to make it K8s/OS scope like `che.infra.kubernetes.server_secure_exposer`?

[the remark this answers is not in the document]
Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: interesting question) On one hand it would be more complex to configure Che if we had several properties for several infras. On the other hand, we don't have anything like this for the Docker infra. 
I think it is not so critical, and I would opt for the 1st option.

[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: OK, I will rename it to `che.server.secure_exposer` and we'll rename it to infraspecific if it will be needed to have different authentication proxy backends for different infrastructures.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+   * with the specified runtime identity.
+   */
+  public SecureServerExposer<T> create(RuntimeIdentity identity) {
+    if (!agentsAuthEnabled) {
+      // return default secure server exposer because no need to protect servers with authentication
+      return new DefaultSecureServerExposer<>(exposerStrategy);
+    }
+
+    switch (serverExposer) {
+      case "jwtproxy":
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: Makes sense, I'll rework it with Map bindings.

[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: Ouch. I missed that `JwtProxySecureServerExposer` is created with factory that uses assisted inject. So, SecureServerExposerFactory could inject something like `Map<String, SecureServerExposerImplFactory>` where `SecureServerExposerImplFactory` is a custom interface 
```
interface SecureServerExposerImplFactory {
  SecureServerExpose create(RuntimeIdentity identity);
}
```
And then there should be corresponding SecureServerExposerImplFactory implementaion of each secure server (now JwtProxy and Default).

Looks like a bit overkill for now.

Do you think I should rework it or it is OK to leave it as it is for now?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: I'm OK to leave the old name. I missed that it is old.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: Maybe we can use a single interface of a factory of a factory instead of the map) This would allow using another implementation instead of JWT one (it can be even customized version of JWT one, depending on the ISV needs).
But yes, if there is no elegant and easy to implement and support way or if it will delay merging of the PR we can totally return to this issue when there would be a need for that.
And it looks like we definitely need to leave the approach as it is implemented now.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: `che.agents.auth_enabled` leaved as it is and `che.agents.secure_server_exposer` is renamed to `che.server.secure_exposer`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: Reworked in the suggested way.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.external.ExternalServerExposerStrategy;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.secure.SecureServerExposer.DefaultSecureServerExposer;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.secure.jwtproxy.JwtProxySecureServerExposerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** @author Sergii Leshchenko */
+public class SecureServerExposerFactory<T extends KubernetesEnvironment> {
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java]: Added

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java (hunk 3 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java]: Added

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxySecureServerExposer.java (hunk 4 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxySecureServerExposer.java]: Added

**Comment by sleshchenko** [comment]
> Is this PR working or just a part of working new security system?

Actually, it requires a few more things to be working:
- excludes will be implemented soon in jwtproxy otherwise liveness probes fail because of token leak. For testing the following workaround was used https://github.com/eclipse/che/commit/fe8ff5fb88d6629e0c024b73bb2e248a9c02d155
- now jwtproxy requires some claims to be present in JWT while they are documented as optional in a specification. So, jwt proxy will be improved not to require optional claims and some of claims will be added to machine token soon. Temporary commit for testing is here https://github.com/eclipse/che/commit/8f7c9759d07646e6572a8cf68a667953bd52c5bd

So, when these two steps will be merged, authentication secure servers with JWT proxy will be working in master branch (but disabled by default).

**Review by garagatyi — APPROVED** [review]
Looks good!

**Comment by sleshchenko** [comment]
> Then I think we should not have release notes in this PR as it would not get into the product as something users can use.

It makes sense. I'll remove release notes section

**Comment by sleshchenko** [comment]
ci-test

**Comment by riuvshin** [comment]
ci-test build report:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/465/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/465//Selenium_tests_report)
[selenium tests report data](http://tests.codenvycorp.com/selenium_test_report_data/10252)
docker image: eclipseche/che-server:10252
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Comment by sleshchenko** [comment]
@eclipse/eclipse-che-qa Could you please review test report?

**Comment by Ohrimenko1988** [comment]
@sleshchenko In the functionality which is covered by selenium tests - new regression was not found.

**Comment by sleshchenko** [comment]
During manual testing on minikube found the following error:
screenshot_20180712_114610

I'll prepare the corresponding changes to reduce service port name.
